google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/apis/CloudApiManager.java

```diff
@@ -69,67 +75,99 @@ static void enableApis(Set<CloudLibrary> libraries, CloudProject cloudProject, P
       return;
     }
 
-    ServiceManagement serviceManagement =
-        GoogleApiClientFactory.getInstance().getServiceManagementClient(user.get().getCredential());
-
-    List<CloudLibrary> enabledApis = Lists.newArrayList();
-    List<String> erroredApiNames = Lists.newArrayList();
-
-    libraries.forEach(
-        library -> {
-          try {
-            serviceManagement
-                .services()
-                .enable(
-                    library.getServiceName(),
-                    new EnableServiceRequest()
-                        .setConsumerId(
-                            String.format(
-                                SERVICE_REQUEST_PROJECT_PATTERN, cloudProject.projectId())))
-                .execute();
-            enabledApis.add(library);
-          } catch (IOException e) {
-            LOG.warn(
-                "Exception occurred attempting to enable API " + library.getName() + " on GCP", e);
-            erroredApiNames.add(library.getName());
-          }
-        });
-
-    if (!erroredApiNames.isEmpty()) {
-      notifyApiEnableError(erroredApiNames, project);
+    List<CloudLibrary> libraryList = new ArrayList<>(libraries);
+    Set<CloudLibrary> enabledApis = Sets.newHashSet();
+    Set<CloudLibrary> erroredApis = Sets.newHashSet();
+
+    for (int i = 0; i < libraryList.size(); i++) {
+      CloudLibrary library = libraryList.get(i);
+
+      try {
+        ProgressIndicator progress = ProgressManager.getInstance().getProgressIndicator();
+
+        if (progress.isCanceled()) {
+          LOG.info("API enablement canceled by user");
+          notifyApiEnableSkipped(Sets.difference(libraries, enabledApis), project);
+          return;
+        }
+
+        setApiEnableProgress(
+            progress,
+            library.getName(),
+            cloudProject.projectName(),
+            (double) i / libraryList.size());
+        enableApi(library, cloudProject, user.get());
+
+        enabledApis.add(library);
+      } catch (IOException e) {
```

Comment:
Does `enableApi` call use same `IOException` to identify I/O and all other (i.e. auth or billing) errors? Or it's expected this call always succeeds if I/O doesn't fail?

Reply:
These use the generic (autogenerated?) apis to manage things on gcp. So yes, I believe that it rolls all exceptions into an IOException.

Reply:
So if this happens, all problematic APIs will be shown in error notification and error stacktraces printed in log, correct? And they can either retry from IDE or go to GCP console to fix.

Reply:
Correct. 

They can retry from the IDE, but there is not yet a good story for this (since currently the enablement is tightly coupled to client library addition). So it won't exactly be very clear how to retry. This goes along with what @elharo brought up - we do plan on having a standalone mechanism for api enablement.